specs/ChainSpec/Consensus.md

```diff
@@ -0,0 +1,219 @@
+
+If a block to be produced will have height `h` and previous block `B'`, the approvals logically included in it must be an `Endorsement` with the hash of `B'` if and only if `h == B'.height + 1`, otherwise it must be a `Skip` with the height of `B'`. See [this section](#approval-condition) below for details on why the endorsements must contain the hash of the previous block, and skips must contain the height.
+
+Note that since each approval that is logically stored in the block is the same for each block producer (except for the `account_id` of the sender and the `signature`), it is redundant to store the full approvals. Instead physically we only store the signatures of the approvals. The specific way they are stored is the following: we first fetch the ordered set of block producers from the current epoch. If the block is on the epoch boundary and also needs to include approvals from the next epoch (see [epoch switches](#epoch-switches)), we add new accounts from the new epoch
+
+```python
+def get_accounts_for_block_ordered(h, prev_block):
+    cur_epoch = get_next_block_epoch(prev_block)
+    next_epoch = get_next_block_next_epoch(prev_block)
+
+    account_ids = get_epoch_block_producers_ordered(cur_epoch)
+    if next_block_needs_approvals_from_next_epoch(prev_block):
+        for account_id in get_epoch_block_producers_ordered(next_epoch):
+            if account_id not in account_ids:
+                account_ids.append(account_id)
+
+    return account_ids
+```
+
+The block then contains a vector of optional signatures of the same or smaller size than the resulting set of `account_ids`, with each element being `None` if the approval for such account is absent, or the signature on the approval message if it is present. It's easy to show that the actual approvals that were signed by the block producers can easily be reconstructed from the information available in the block, and thus the signatures can be verified. If the vector of signatures is shorter than the length of `account_ids`, the remaining signatures are assumed to be `None`.
+
+## Messages
+
+On receival of the approval message the participant just stores it in the collection of approval messages.
+
+```python
+def on_approval(self, approval):
+    self.approvals.append(approval)
+```
+
+Whenever a participant receives a block, the operations relevant to the consensus include updating the `head` and initiating a timer to start sending the approvals on the block to the block producers at the consecutive `target_height`s. The timer delays depend on the height of the last final block, so that information is also persisted.
+
+```python
+def on_block(self, block):
+    header = block.header
```

Comment:
does on_approval and on_block all comes after "check approval/block valid & not repetitive"?

Reply:
yes